hummingbot/cli/utils/stop_loss_tracker.py

```diff
@@ -0,0 +1,96 @@
+import logging
+import asyncio
+from typing import (
+    List,
+    Dict,
+    Callable,
+    Optional,
+)
+
+from hummingbot.data_feed.data_feed_base import DataFeedBase
+from hummingbot.cli.config.global_config_map import global_config_map
+from wings.market.market_base import MarketBase
+
+
+class StopLossTracker:
+    slc_logger: Optional[logging.Logger] = None
+
+    @classmethod
+    def logger(cls) -> logging.Logger:
+        if cls.slc_logger is None:
+            cls.slc_logger = logging.getLogger(__name__)
+        return cls.slc_logger
+
+    def __init__(self,
+                 data_feed: DataFeedBase,
+                 assets: List[str],
+                 markets: List[MarketBase],
+                 stop_handler: Callable):
+        self._data_feed: DataFeedBase = data_feed
+        self._markets: List[MarketBase] = markets
+        self._assets: List[str] = assets
+        self._stop_handler: Callable = stop_handler
+        self._stop_loss_pct = global_config_map.get("stop_loss_pct").value
+        self._stop_loss_price_type = global_config_map.get("stop_loss_price_type").value
+        self._stop_loss_base_token = global_config_map.get("stop_loss_base_token").value
+
+        self._starting_balances: Dict[str, float] = {}
+        self._starting_prices: Dict[str, float] = {}
+        self._current_pnl = 0.0
+        self._started = False
+        self._update_interval = 10.0
+        self._check_stop_loss_task: Optional[asyncio.Task] = None
+
+    @property
+    def current_pnl(self):
+        return self._current_pnl
+
+    def get_balances(self) -> Dict[str, float]:
+        balance_dict: Dict[str, float] = {}
+        for asset in self._assets:
+            balance_dict[asset] = 0.0
+            for market in self._markets:
+                balance_dict[asset] += market.get_balance(asset)
+        return balance_dict
+
+    async def stop_loss_loop(self):
+        while True:
+            self._current_pnl = self.calculate_profit_loss_pct(self._stop_loss_price_type)
```

Comment:
All `while True` loops should have a `try... except...` block in case they have any errors. Two reasons:

 1. We don't want background processes to fail silently - at the minimum they should record an error log if they have an unrecoverable error.
 2. If the errors are recoverable, we normally want to just record the error log and continue - rather than leave the task killed.

Reply:
fixed